CodeiumVS/Utilities/Extensions.cs

```diff
@@ -1,24 +1,59 @@
-﻿using Microsoft.VisualStudio.Text.Editor;
+﻿using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
 using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.Utilities;
 using System.Collections.Generic;
+using System.Diagnostics;
 
 namespace CodeiumVS.Utilities;
 
-internal abstract class TextViewExtension<T> where T : class
+internal abstract class PropertyOwnerExtension<OwnerType, ExtensionType> : IDisposable
+    where OwnerType     : IPropertyOwner
+    where ExtensionType : class
 {
-    protected ITextView _hostView { get; set; }
+    protected bool _disposed = false;
+    protected readonly OwnerType _owner;
 
-    private static readonly Dictionary<ITextView, T> Instances = [];
+    public PropertyOwnerExtension(OwnerType owner)
+    {
+        _owner = owner;
+        _owner.Properties.AddProperty(typeof(ExtensionType), this as ExtensionType);
+    }
+
+    public static ExtensionType? GetInstance(OwnerType owner)
+    {
+        return owner.Properties.TryGetProperty(typeof(ExtensionType), out ExtensionType instance) ? instance : null;
```

Comment:
Switched to this instead of a Dictionary, the old one had race condition, this is the standard that VS uses internally.

Reply:
Be aware that `TryGetProperty` uses the locking mechanism so it's a good practice not to call it too often.